plane/src/controller/proxy.rs

```diff
@@ -19,31 +22,135 @@ use std::net::{IpAddr, SocketAddr};
 use tokio::select;
 use valuable::Valuable;
 
-pub async fn handle_message_from_proxy(
-    message: MessageFromProxy,
+pub async fn handle_route_info_request(
+    token: BearerToken,
     controller: &Controller,
     socket: &mut TypedSocket<MessageToProxy>,
-    cluster: &ClusterName,
-    node_id: NodeId,
 ) {
-    match message {
-        MessageFromProxy::RouteInfoRequest(RouteInfoRequest { token }) => {
-            let route_info = match controller.db.backend().route_info_for_token(&token).await {
-                Ok(route_info) => route_info,
-                Err(err) => {
-                    tracing::error!(?err, "Error getting route info");
-                    return;
-                }
+    match controller.db.backend().route_info_for_token(&token).await {
+        // When a proxy requests a route, either:
+        // 1. The route is ready, and we can send it back immediately.
+        // 2. The route is not ready, and we need to wait for it to become ready.
+        // 3. The route does not exist or has already terminated, and we can send back a `None`.
+        Ok(RouteInfoResult::Ready(route_info)) => {
+            let response = RouteInfoResponse {
+                token,
+                route_info: Some(route_info),
             };
+            if let Err(err) = socket
+                .send(MessageToProxy::RouteInfoResponse(response))
+                .await
+            {
+                tracing::error!(?err, "Error sending route info response to proxy.");
+            }
+        }
+        Ok(RouteInfoResult::Waiting(partial_route_info)) => {
+            let backend_id = partial_route_info.backend_id.clone();
+            let mut sub: Subscription<BackendState> =
+                controller.db.subscribe_with_key(backend_id.as_str());
+
+            // There is a race condition if the status updated between when our last query hit and when we started the
+            // subscription. It's a bit hacky, but for now we will just issue the query again.
+            // We can't start the subscription first to avoid repeating the query, because we need to know the backend
+            // ID to start the subscription.
+            if let Ok(RouteInfoResult::Ready(route_info)) =
+                controller.db.backend().route_info_for_token(&token).await
+            {
+                let response = RouteInfoResponse {
+                    token,
+                    route_info: Some(route_info),
+                };
+                if let Err(err) = socket
+                    .send(MessageToProxy::RouteInfoResponse(response))
+                    .await
+                {
+                    tracing::error!(?err, "Error sending route info response to proxy.");
+                }
+                return;
+            }
+
+            let socket = socket.sender(MessageToProxy::RouteInfoResponse);
+            tokio::spawn(async move {
+                while let Some(result) = sub.next().await {
```

Comment:
I think it would be a good idea to have a timeout here to prevent hanging forever and creating a memory leak (an issue we've had to track down and fix elsewhere)